swabbie-core/src/main/kotlin/com/netflix/spinnaker/swabbie/notifications/NotificationSender.kt

```diff
@@ -0,0 +1,165 @@
+
+  /**
+   * Periodically sends notifications to resolved resource owners
+   * The frequency of this scheduled action is cron based and configurable
+   * Defaults to every day at 9am.
+   */
+  @Scheduled(cron = "\${swabbie.notification.cron.schedule:0 0 9 1/1 * ?}")
+  fun sendNotifications() {
+    if (!isUp()) {
+      return
+    }
+
+    lockingService.acquireLock(lockOptions) {
+      val resources = resourceTrackingRepository.getMarkedResources()
+        .filter {
+          it.notificationInfo == null
+        }
+
+      groupRelatedResourcesThenNotify(resources)
+    }
+  }
+
+  /**
+   * @return Marked resources to notify on
+   */
+  private fun groupRelatedResourcesThenNotify(markedResources: List<MarkedResource>) {
+    val now = Instant.now(clock)
+    val tasks = mutableListOf<NotificationTask>()
+    do {
+      val notificationTask: NotificationTask? = notificationQueue.pop()
+      if (notificationTask == null) {
+        log.debug("Nothing to notify. Skipping...")
+        return
+      }
+
+      tasks.add(notificationTask)
+    } while (!notificationQueue.isEmpty())
+
+    tasks.forEach { task ->
+      val byResourceOwner = markedResources
+        .filter {
+          it.resourceType == task.resourceType
+        }.groupBy {
+          it.resourceOwner
+        }
+
+      byResourceOwner.forEach { (owner, list) ->
+        val resourcesAndConfigurations: List<Pair<MarkedResource, WorkConfiguration>> = list.toResourceAndConfigurationPairs()
+        val result = notifier.notify(
+          Notifier.Envelope(
+            recipient = owner,
+            resources = resourcesAndConfigurations
+          )
+        )
+
+        if (result.success) {
+          log.debug("Sent notification to {} for {}:{}",
+            owner, task.resourceType, markedResources.size)
+        }
+
+        resourcesAndConfigurations.forEach { (resource, workConfiguration) ->
+          if (result.success) {
+            // Update resource's notification info
+            resource
+              .withNotificationInfo(
+                NotificationInfo(result.recipient, result.notificationType.name, now.toEpochMilli())
+              )
+              .withAdditionalTimeForDeletion(
+                Instant.ofEpochMilli(resource.markTs!!).until(now, ChronoUnit.MILLIS)
+              ).also {
+                resourceTrackingRepository.upsert(it)
+                applicationEventPublisher.publishEvent(OwnerNotifiedEvent(it, workConfiguration))
+              }
+
+            registry.counter(notificationsId.withTags("result", "success")).increment()
+          } else {
+            // unmark this resource & let it go through the mark cycle later
+            registry.counter(notificationsId.withTags("result", "failure")).increment()
+            resourceTrackingRepository.remove(resource)
+            applicationEventPublisher.publishEvent(UnMarkResourceEvent(resource, workConfiguration))
```

Comment:
I think this should be a different type of event, so that we can log metrics independently for things that fail to notify. I don't want to lump everything into the "unmark" bucket.

Reply:
There is a counter right above tracking notification failures.

Thinking about this more, the resource would get picked up when we notify on the same resource type. So no need to remove and unmark. It could use a debug level log here as well.

Reply:
ahh I read so well. I totally see that now :)